huxley/api/permissions.py

```diff
@@ -67,56 +67,78 @@ def has_permission(self, request, view):
         if request.user.is_superuser:
             return True
 
-        delegate_id = view.kwargs.get('pk', None)
-        delegate = Delegate.objects.get(id=delegate_id)
-        user = request.user
-
-        return (user.is_authenticated() and user.is_advisor() and
-                user.school.id == delegate.school.id)
+        if request.method in permissions.SAFE_METHODS:
+            return user_is_chair(request, view) or user_is_advisor(request, view)
 
+        return False
+        
 
-class AssignmentListPermission(permissions.BasePermission):
+class DelegateDetailPermission(permissions.BasePermission):
+    '''Accept requests to create, get, and update a delegate from the
+       superuser, the advisor of the shcool of the delegate, or the chair 
+       of the committee of the delegate. Chairs should not be able to
+       delete delegates.'''
 
     def has_permission(self, request, view):
         if request.user.is_superuser:
             return True
 
-        if request.method in permissions.SAFE_METHODS:
-            return user_is_advisor(request, view)
+        delegate_id = view.kwargs.get('pk', None)
+        delegate = Delegate.objects.get(id=delegate_id)
+        user = request.user
+        if not user.is_authenticated():
+            return False
+
+        if user.is_advisor():
+            return user.school.id == delegate.school.id
+
+        if user.is_chair() and request.method != 'DELETE':
+            return user.committee.id == delegate.assignment.committee.id
 
         return False
 
 
 class DelegateListPermission(permissions.BasePermission):
     '''Accept requests to create, get and update delegates in bulk from
-       the superuser and from the advisor of the school of the delegates.'''
+       the superuser and from the advisor of the school of the delegates.
+       Accept requests to get and update delegates from the chair of the
+       committee of the delegates.'''
 
     def has_permission(self, request, view):
         user = request.user
         if user.is_superuser:
             return True
 
+        if not user.is_authenticated():
+            return False
+
         method = request.method
         if method in permissions.SAFE_METHODS:
-            return user_is_advisor(request, view)
-
-        if method in ('POST', 'PUT', 'PATCH'):
-            school_id = user.is_authenticated() and user.school_id
-            if not school_id:
-                return False
-
-            if method == 'POST':
-                return int(request.data['school']) == school_id
-            else:
-                delegate_ids = [delegate['id'] for delegate in request.data]
-                return not Delegate.objects.filter(id__in=delegate_ids).exclude(school_id=school_id).exists()
+            return user_is_chair(request, view) or user_is_advisor(request, view)
+
+        if method == 'POST':
+            school_id = user.school_id
+            return (user.is_advisor() and school_id and 
```

Comment:
I think we can write the `user_is_advisor` function to be a bit more general by accepting a `school_id` parameter, so that we don't have to do all these checks everywhere we need to check for an advisor.